test/sync_tests/conflicting.test.js

```diff
@@ -0,0 +1,763 @@
+/* eslint-disable no-unused-expressions */
+/* eslint-disable no-undef */
+import * as Factory from '../lib/factory.js';
+chai.use(chaiAsPromised);
+const expect = chai.expect;
+
+describe('online conflict handling', function () {
+  this.timeout(Factory.TestTimeout);
+  const BASE_ITEM_COUNT = 2; /** Default items key, user preferences */
+
+  const syncOptions = {
+    checkIntegrity: true,
+    awaitAll: true,
+  };
+
+  beforeEach(async function () {
+    localStorage.clear();
+    this.expectedItemCount = BASE_ITEM_COUNT;
+    this.application = await Factory.createInitAppWithRandNamespace();
+    this.email = Uuid.GenerateUuidSynchronously();
+    this.password = Uuid.GenerateUuidSynchronously();
+    await Factory.registerUserToApplication({
+      application: this.application,
+      email: this.email,
+      password: this.password,
+    });
+  });
+
+  afterEach(async function () {
+    expect(this.application.syncService.isOutOfSync()).to.equal(false);
+    const items = this.application.itemManager.items;
+    expect(items.length).to.equal(this.expectedItemCount);
+    const rawPayloads = await this.application.storageService.getAllRawPayloads();
+    expect(rawPayloads.length).to.equal(this.expectedItemCount);
+    await this.application.deinit();
+    localStorage.clear();
+  });
+
+  function createDirtyPayload(contentType) {
+    const params = {
+      uuid: Uuid.GenerateUuidSynchronously(),
+      content_type: contentType,
+      content: {
+        foo: 'bar',
+      },
+    };
+    const payload = CreateMaxPayloadFromAnyObject(params, {
+      dirty: true,
+      dirtiedDate: new Date(),
+    });
+    return payload;
+  }
+
+  it('components should not be duplicated under any circumstances', async function () {
+    const payload = createDirtyPayload(ContentType.Component);
+    const item = await this.application.itemManager.emitItemFromPayload(
+      payload,
+      PayloadSource.LocalChanged
+    );
+    this.expectedItemCount++;
+    await this.application.syncService.sync(syncOptions);
+    /** First modify the item without saving so that
+     * our local contents digress from the server's */
+    await this.application.changeItem(item.uuid, (mutator) => {
+      mutator.content.foo = `${Math.random()}`;
+    });
+    await this.application.changeAndSaveItem(
```

Comment:
How does `changeAndSaveItem()` not override what `changeItem()` did? I don't understand how the conflict is created…

Reply:
changeItem changes the item but does not sync. changeAndSaveItem changes and syncs.

The goal is to change the value locally before we sync so that by the time conflicts are processed, our previous revision's value has a different value from the server.

Reply:
Since changeAndSaveItem is essentially calling changeItem and then syncing, and you're doing it in the same item, I read this as changing the item twice locally and then syncing, which I wouldn't expect to create a conflict.